File: frontend/lib/agent/device-tools.ts
/**
 * Device-mode tool routing — Arduino vs Raspberry Pi
 */

import type { DeviceMode } from '@/lib/arduino-studio/types';
import type { ToolName, ToolDefinition } from '@/lib/arduino-studio/tool-registry';
import { TOOL_REGISTRY } from '@/lib/arduino-studio/tool-registry';

/** Remap device-sensitive tools to the correct action for current Studio mode */
export function resolveToolForDevice(tool: string, deviceMode: DeviceMode): ToolName {
  const t = tool as ToolName;
  if (deviceMode === 'raspberry-pi') {
    if (t === 'GENERATE_ARDUINO') return 'GENERATE_RPI';
    if (t === 'VERIFY_ARDUINO') return 'VERIFY_PYTHON';
  } else {
    if (t === 'GENERATE_RPI') return 'GENERATE_ARDUINO';
    if (t === 'VERIFY_PYTHON') return 'VERIFY_ARDUINO';
  }
  return t;
}

/** Tools exposed to the LLM for the active device mode */
export function getToolsForDeviceMode(deviceMode: DeviceMode): ToolDefinition[] {
  const exclude: Set<ToolName> =
    deviceMode === 'raspberry-pi'
      ? new Set(['GENERATE_ARDUINO', 'VERIFY_ARDUINO'])
      : new Set(['GENERATE_RPI', 'VERIFY_PYTHON']);

  return TOOL_REGISTRY.filter(t => !exclude.has(t.name));
}

export function buildDeviceSystemSuffix(deviceMode: DeviceMode): string {
  if (deviceMode === 'raspberry-pi') {
    return `## Active Studio Mode: Raspberry Pi

You are building **Raspberry Pi Python/GPIO** projects.
- Use **GENERATE_RPI** (not GENERATE_ARDUINO) for new code
- Use **VERIFY_PYTHON** (not VERIFY_ARDUINO) for verification
- Use BCM GPIO numbering in explanations
- Default entry file: \`main.py\``;
  }
  return `## Active Studio Mode: Arduino

You are building **Arduino C++** embedded projects.
- Use **GENERATE_ARDUINO** (not GENERATE_RPI) for new code
- Use **VERIFY_ARDUINO** (not VERIFY_PYTHON) for verification
- Default entry file: \`sketch.ino\``;
}

/** Default generate tool for slash commands */
export function defaultGenerateTool(deviceMode: DeviceMode): ToolName {
  return deviceMode === 'raspberry-pi' ? 'GENERATE_RPI' : 'GENERATE_ARDUINO';
}

export function defaultVerifyTool(deviceMode: DeviceMode): ToolName {
  return deviceMode === 'raspberry-pi' ? 'VERIFY_PYTHON' : 'VERIFY_ARDUINO';
}

/** Build a direct tool_call message the agent loop can execute */
export function buildSlashToolMessage(
  tool: ToolName,
  userPrompt: string,
  summary: string
): string {
  return `\`\`\`tool_call
{"tool": "${tool}", "args": {"_summary": "${summary}", "prompt": ${JSON.stringify(userPrompt)}}}
\`\`\``;
}

File: backend/app/agent/device_tools.py
"""Device-mode tool routing — port of frontend/lib/agent/device-tools.ts."""

from __future__ import annotations

from typing import Literal

DeviceMode = Literal["arduino", "raspberry-pi"]

# Tools excluded from each mode
_EXCLUDE_FROM_RPI = {"GENERATE_ARDUINO", "VERIFY_ARDUINO"}
_EXCLUDE_FROM_ARDUINO = {"GENERATE_RPI", "VERIFY_PYTHON"}

# All tool names in registry order
_ALL_TOOLS = [
    "GENERATE_ARDUINO",
    "GENERATE_RPI",
    "VERIFY_ARDUINO",
    "VERIFY_PYTHON",
    "DERIVE_WIRING",
    "OPTIMIZE_CODE",
    "APPLY_CHANGES",
    "SEND_MESSAGE",
    "WEB_SEARCH",
    "FETCH_PROMPTS",
    "READ_FILE",
    "LIST_FILES",
    "WRITE_FILE",
]


def resolve_tool_for_device(tool: str, device_mode: DeviceMode) -> str:
    """Return the effective tool name for this device mode.

    If GENERATE_ARDUINO is called in RPi mode → GENERATE_RPI (and vice versa).
    If VERIFY_ARDUINO is called in RPi mode → VERIFY_PYTHON (and vice versa).
    """
    if device_mode == "raspberry-pi":
        if tool == "GENERATE_ARDUINO":
            return "GENERATE_RPI"
        if tool == "VERIFY_ARDUINO":
            return "VERIFY_PYTHON"
    else:
        if tool == "GENERATE_RPI":
            return "GENERATE_ARDUINO"
        if tool == "VERIFY_PYTHON":
            return "VERIFY_ARDUINO"
    return tool


def get_tools_for_device_mode(device_mode: DeviceMode) -> list[str]:
    """Return the tool names available for the given device mode."""
    exclude = _EXCLUDE_FROM_RPI if device_mode == "raspberry-pi" else _EXCLUDE_FROM_ARDUINO
    return [t for t in _ALL_TOOLS if t not in exclude]


def default_generate_tool(device_mode: DeviceMode) -> str:
    return "GENERATE_RPI" if device_mode == "raspberry-pi" else "GENERATE_ARDUINO"


def default_verify_tool(device_mode: DeviceMode) -> str:
    return "VERIFY_PYTHON" if device_mode == "raspberry-pi" else "VERIFY_ARDUINO"


def build_device_system_suffix(device_mode: DeviceMode) -> str:
    if device_mode == "raspberry-pi":
        return (
            "\n\n## Mode: Raspberry Pi\n"
            "You are working in Raspberry Pi mode. Generate Python 3 code using RPi.GPIO or gpiozero. "
            "Use BCM pin numbering. Produce wiring manifests with Raspberry Pi board types. "
            "Use GENERATE_RPI for code generation and VERIFY_PYTHON for verification."
        )
    return (
        "\n\n## Mode: Arduino\n"
        "You are working in Arduino mode. Generate Arduino C++ sketches. "
        "Use pin numbers valid for Arduino Uno/Nano/Mega/ESP32. "
        "Use GENERATE_ARDUINO for code generation and VERIFY_ARDUINO for verification."
    )
